### EX: Forward Propagation from Scratch (Mines vs. Rocks)

In this exercise, we will pull back the curtain on how a neural network makes a decision. Instead of relying on high-level libraries like Keras or Scikit-Learn to do the heavy lifting, we will execute a **Forward Propagation** pass manually using NumPy. By explicitly defining the matrices for our weights and biases, we can trace exactly how raw sensor data is mathematically transformed into a final prediction.

#### Steps Performed

* **Define Activation Functions:** We define two core mathematical functions: `relu` (Rectified Linear Unit) for our hidden layer to introduce non-linearity, and `sigmoid` for our final output layer to squash our result into a probability between **0.0** and **1.0**.

* **Simulate Data:** We create a single mock sonar return containing 3 features (e.g., Frequency, Ping Duration, Amplitude).

* **Define Network Architecture:** We manually set up the weight matrices and bias arrays for a multi-layer perceptron. The network maps **3 Inputs** $\rightarrow$ **4 Hidden Neurons** $\rightarrow$ **1 Output Neuron**.

* **Execute Forward Propagation:** We use NumPy's dot product (`np.dot`) to calculate the weighted sum of the inputs and weights, add the bias, and pass the result through the activation functions for each layer sequentially.

* **Thresholding and Output:** We print the exact values at each step of the calculation and apply a standard **0.5** threshold to classify the final probability as either a "Subsurface Mine" or a "Harmless Rock".




In [ ]:
#Only run this cell after selecting your kernel
!python.exe -m pip install --upgrade pip
!pip install numpy matplotlib scikit-learn pandas tensorflow

In [1]:
# EX: Forward Propagation (Mines vs. Rocks Sonar Returns)
# In this lab, we will execute forward propagation manually using NumPy to understand the underlying matrix math.

import numpy as np

# 1. Define the Activation Functions
def relu(z):
    return np.maximum(0, z)

def sigmoid(z):
    return 1 / (1 + np.exp(-z))

# 2. Simulate Sonar Input Data (1 Sample, 3 Features: Frequency, Ping Duration, Amplitude)
X = np.array([[0.8, 1.2, -0.5]])

# 3. Define the Network Architecture (3 Inputs -> 4 Hidden Neurons -> 1 Output)
# Weights and biases for Hidden Layer 1
W1 = np.array([
    [ 0.2, -0.1,  0.4,  0.8],
    [ 0.5,  0.6, -0.3, -0.2],
    [-0.4,  0.3,  0.1,  0.5]
])
b1 = np.array([[0.1, -0.2, 0.0, 0.1]])

# Weights and biases for Output Layer
W2 = np.array([
    [ 0.6],
    [-0.5],
    [ 0.8],
    [-0.3]
])
b2 = np.array([[-0.1]])

# 4. Execute Forward Propagation
# Hidden Layer: Weighted sum (Z1) then ReLU (A1)
Z1 = np.dot(X, W1) + b1
A1 = relu(Z1)

# Output Layer: Weighted sum (Z2) then Sigmoid (A2)
Z2 = np.dot(A1, W2) + b2
A2 = sigmoid(Z2)

print("--- Forward Propagation Results ---")
print(f"Input Features (X): {X}")
print(f"Hidden Layer Pre-Activation (Z1): {Z1}")
print(f"Hidden Layer Post-ReLU (A1): {A1}")
print(f"Output Pre-Activation (Z2): {Z2}")
print(f"Final Prediction Probability (A2): {A2[0][0]:.4f}")

# Threshold logic
if A2[0][0] > 0.5:
    print("Classification: Subsurface Mine Detected")
else:
    print("Classification: Harmless Rock")

--- Forward Propagation Results ---
Input Features (X): [[ 0.8  1.2 -0.5]]
Hidden Layer Pre-Activation (Z1): [[ 1.06  0.29 -0.09  0.25]]
Hidden Layer Post-ReLU (A1): [[1.06 0.29 0.   0.25]]
Output Pre-Activation (Z2): [[0.316]]
Final Prediction Probability (A2): 0.5783
Classification: Subsurface Mine Detected


#### Interpreting the Results

By tracing the matrix math manually, we can see exactly how the network processes the raw sensor telemetry to arrive at a tactical decision:

1. **The Hidden Layer Weighted Sum ($Z_1$):** Our 3 input features are multiplied by the $3 \times 4$ weight matrix $W_1$, and the biases $b_1$ are added. This outputs an array of 4 numbers: **[1.06, 0.29, -0.09, 0.25]**. These represent the raw, pre-activation "signals" captured by the 4 neurons in our hidden layer.

2. **The ReLU Activation ($A_1$):** Notice what happens to the third value (**-0.09**) when passed through the ReLU function ($\max(0, z)$). It is immediately squashed to **0.0**. This is the network discarding unhelpful or negative signal noise, introducing the crucial non-linearity that allows neural networks to solve complex problems.

3. **The Output Layer ($Z_2$ and $A_2$):** The activated signals from the hidden layer are fed into the final output neuron. They are multiplied by $W_2$ and added to $b_2$, yielding a raw output of **0.316**. Because 0.316 is just an arbitrary floating-point number, we pass it through the **Sigmoid** activation function, translating it into a highly interpretable probability: **0.5783**.

**The Tactical Takeaway:** Because the final probability is **~57.8%**, it crosses our standard **0.5** decision threshold. The network successfully flags this specific sonar return as a subsurface mine based entirely on the learned weights and non-linear transformations applied to the raw signal.